Source/Core/Core/PowerPC/GDBThread.cpp

```diff
@@ -0,0 +1,1296 @@
+      break;
+    case 68:
+      CTR = ReadHexBE32(bufptr);
+      break;
+    case 69:
+      XER = ReadHexBE32(bufptr);
+      break;
+    case 70:
+      // do nothing, we don't have MQ
+      break;
+    case 71:
+      FPSCR.Hex = ReadHexBE32(bufptr);
+      break;
+    default:
+      return GDBReply("E01");
+      break;
+    }
+  }
+
+  GDBReply("OK");
+}
+
+void GDBThread::GDBHandleQuery()
+{
+  if (!m_connected)
+    return;
+
+  INFO_LOG(GDB_THREAD, "gdb: query '%s'\n", m_cmd_bfr + 1);
+  GDBAck();
+  GDBReply("");
+}
+
+void GDBThread::GDBStep()
+{
+  if (!m_connected)
+    return;
+
+  GDBAck();
+
+  if (CPU::State::Stepping != CPU::GetState())
+    return;
+
+  PowerPC::breakpoints.ClearAllTemporary();
+  JitInterface::InvalidateICache(PC, 4, true);
+
+  Common::Event sync_event;
+
+  sync_event.Reset();
+  CPU::StepOpcode(&sync_event);
+
+  sync_event.Wait();
+
+  GDBSignal((u32)SignalTypes::Trap);
+}
+
+void GDBThread::GDBAddBreakpoint()
+{
+  if (!m_connected)
+    return;
+
+  u32 type = 0;
+  u32 i = 1;
+
+  GDBAck();
+
+  while (m_cmd_bfr[i] != ',')
+    type = (type << 4) | HexToChar(m_cmd_bfr[i++]);
+  i++;
+
+  GDBBPType breakpointType = GDBBPType::None;
+  switch (type)
+  {
+  case 0:
+  case 1:
+    breakpointType = GDBBPType::Execute;
+    break;
+  case 2:
+    breakpointType = GDBBPType::Write;
+    break;
+  case 3:
+    breakpointType = GDBBPType::Read;
+    break;
+  case 4:
+    breakpointType = GDBBPType::Access;
+    break;
+  default:
+    return GDBReply("E01");
+  }
+
+  u32 addr = 0;
+  u32 len = 0;
+
+  while (m_cmd_bfr[i] != ',')
+    addr = (addr << 4) | HexToChar(m_cmd_bfr[i++]);
+  i++;
+
+  while (i < m_cmd_len)
+    len = (len << 4) | HexToChar(m_cmd_bfr[i++]);
+
+  AddBreakpoint(breakpointType, addr, len);
+  GDBReply("OK");
+}
+
+void GDBThread::GDBRemoveBreakpoint()
+{
+  if (!m_connected)
+    return;
+
+  u32 type = 0;
+  u32 i = 1;
+
+  GDBAck();
+
+  while (m_cmd_bfr[i] != ',')
+    type = (type << 4) | HexToChar(m_cmd_bfr[i++]);
+  i++;
+
+  GDBBPType breakpointType = GDBBPType::None;
+  switch (type)
+  {
+  case 0:
+  case 1:
+    breakpointType = GDBBPType::Execute;
+    break;
+  case 2:
+    breakpointType = GDBBPType::Write;
+    break;
+  case 3:
+    breakpointType = GDBBPType::Read;
+    break;
+  case 4:
+    breakpointType = GDBBPType::Access;
+    break;
+  default:
+    return GDBReply("E01");
+  }
+
+  u32 addr = 0;
+  u32 len = 0;
+
+  while (m_cmd_bfr[i] != ',')
+    addr = (addr << 4) | HexToChar(m_cmd_bfr[i++]);
+  i++;
+
+  while (i < m_cmd_len)
+    len = (len << 4) | HexToChar(m_cmd_bfr[i++]);
+
+  RemoveBreakpoint(breakpointType, addr, len);
+  GDBReply("OK");
+}
+
+void GDBThread::GDBPause()
+{
+  if (!m_connected)
+    return;
+
+  GDBAck();
+
+  if (CPU::State::Stepping != CPU::GetState())
+  {
+    CPU::EnableStepping(true);
+
+    UpdateState(CPU::GetState());
+  }
+}
+
+void GDBThread::GDBSignal(u32 s, u64 addr, MemCheckCondition cond)
```

Comment:
`s` and `m_last_signal` should likely be a `SignalType`. This would remove the need for casting at every callsite.